llvm/include/llvm/IR/VectorUtils.h

```diff
@@ -0,0 +1,74 @@
+//===----------- VectorUtils.h - Vector type utility functions -*- C++ -*-====//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_IR_VECTORUTILS_H
+#define LLVM_IR_VECTORUTILS_H
+
+#include "llvm/IR/DerivedTypes.h"
+#include "llvm/IR/StructWideningUtils.h"
+
+namespace llvm {
+
+/// A helper function for converting Scalar types to vector types. If
+/// the incoming type is void, we return void. If the EC represents a
+/// scalar, we return the scalar type.
+inline Type *ToVectorTy(Type *Scalar, ElementCount EC) {
+  if (Scalar->isVoidTy() || Scalar->isMetadataTy() || EC.isScalar())
+    return Scalar;
+  return VectorType::get(Scalar, EC);
+}
+
+inline Type *ToVectorTy(Type *Scalar, unsigned VF) {
+  return ToVectorTy(Scalar, ElementCount::getFixed(VF));
+}
+
+/// A helper for converting to wider (vector) types. For scalar types, this is
+/// equivalent to calling `ToVectorTy`. For struct types, this returns a new
+/// struct where each element type has been widened to a vector type. Note: Only
+/// unpacked literal struct types are supported.
+inline Type *ToWideTy(Type *Ty, ElementCount EC) {
+  if (StructType *StructTy = dyn_cast<StructType>(Ty))
+    return ToWideStructTy(StructTy, EC);
+  return ToVectorTy(Ty, EC);
+}
+
+/// A helper for converting wide types to narrow (non-vector) types. For vector
+/// types, this is equivalent to calling .getScalarType(). For struct types,
+/// this returns a new struct where each element type has been converted to a
+/// scalar type. Note: Only unpacked literal struct types are supported.
+inline Type *ToNarrowTy(Type *Ty) {
+  if (StructType *StructTy = dyn_cast<StructType>(Ty))
+    return ToNarrowStructTy(StructTy);
+  return Ty->getScalarType();
+}
+
+/// Returns true if `Ty` is a vector type or a struct of vector types where all
+/// vector types share the same VF.
+inline bool isWideTy(Type *Ty) {
```

Comment:
Could we use a more concrete name instead of WideTy given that this is a now placed in `IR` to avoid confusion with other types we use widening terminology for (e.g. widening integer types of indications)

Reply:
Do you like any of these: `VectorizedTy` and `ScalarizedTy`, `WideVectorLikeTy` and `NarrowScalarLikeTy`, `VectorLikeTy` and `ScalarLikeTy`? Naming is tricky.

Reply:
Think `isVectorizedTy` / `ToVectorizedTy` / `ToScalarizedTy` may be the clearest?

Reply:
Right, I've gone with `toVectorizedTy` / `toScalarizedTy`. I think that's the best option, as it refers to what we're doing to the type (vectorizing or scalarizing it), rather than what the result type is.

Reply:
Agreed that naming is tricky :) 

Looking at `toVectorizedTy` and its documentation, something like`toVectorTyOrStructOfVectorsTy`/`toScalarTyOrStructOfScalarsTy` may be slightly more descriptive at the cost of being a bit more verbose though?

Reply:
I honestly don't find such verbose names very helpful, as the names have to lie (to avoid being too verbose), which means you still need to read the docs. I also personally find them a little hard to read (but that's probably just a me thing). 

I've updated the docs of `toVectorizedTy` to include make it clear there's cases where it returns neither a vector type nor a struct type.

Reply:
For what it's worth I'm working on a use case for LoopVectorize to vectorise and/or interleave loops that already contain some vector code and one of the issues I hit are the terms `scalar` and `vector` which do not strictly apply. Something simple like `toVectorizedTy` would be an extensible solution that would cover my need as well rather than trying to be overly verbose.